### DISCUSSION TITLE
Calling `.merge` or `.extend` on recursive types

### DISCUSSION THREAD

**Discussion by celestine-stripe** [colinhacks/zod#2984] [Q&A] [answered]
First off, really impressive project! 

So my use case is probably not a great fit for `zod`, but I'd like to try anyway. I'm struggling with trying to define a schema that involves deeply recursive types. Here's a stripped down example, following the example from the [guide](https://github.com/colinhacks/zod#recursive-types): 

```typescript
const _HashSpecBase = z.object({
  type: z.ZodLiteral('object')
});

type _HashSpecType = z.infer<typeof _HashSpecBase> & {
  value_type?: string | z.infer<typeof ArraySpec>;
};

const HashSpec: z.ZodType<_HashSpecType> = _HashSpecBase.merge(
  z.object({
    value_type: z.lazy(() => z.union([z.string(), ArraySpec]).optional())
  })
);

const _ArraySpecBase = z.object({
  type: z.ZodLiteral('array')
});

type _ArraySpecType = z.infer<typeof _ArraySpecBase> & {
  contains?: string | number | z.infer<typeof HashSpec>;
};

const ArraySpec: z.ZodType<_ArraySpecType> = _ArraySpecBase.merge(
  z.object({
    contains: z.lazy(() => z.union([z.string(), z.number(), HashSpec]).optional())
  })
);
```
The above works fine on its own. Where I run into problems is when I want to call `.extend` or `.merge` on `HashSpec` or `ArraySpec`, for example:
```typescript
const MapSpec = HashSpec.merge(
  z.object({
    max_keys: z.number().optional()
  })
);
```
The above will result in the error `Property 'merge' does not exist on type 'ZodType<_HashSpecType, ZodTypeDef, _HashSpecType>'`, which makes sense given the type hint on `HashSpec`. I've looked at https://github.com/colinhacks/zod/discussions/2255 and https://github.com/colinhacks/zod/discussions/1870 and a few other discussions, but haven't been able to determine a solution that would allow me to call `.extend`/`.merge` on recursive types. So I just wanted to sanity check if I was missing something, or if I should just accept not being able to call `HashSpec.merge` like in the example above, and instead accept some level of duplication with something like:
```typescript
const MapSpec = _HashSpecBase.merge(
z.object({
    value_type: z.lazy(() => z.union([z.string(), ArraySpec]).optional()),
    max_keys: z.number().optional()
  })
);
```
Any advice appreciated, thank you!

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const _HashSpecBase = z.object( {
    type: z.literal( 'object' )
} )

type _HashSpecType = z.infer<typeof _HashSpecBase> & {
    value_type?: string | z.infer<typeof ArraySpec>
}

const hasValueType = {
    value_type: z.lazy( () => z.union( [ z.string(), ArraySpec ] ).optional() )
}

const HashSpec: z.ZodType<_HashSpecType> = _HashSpecBase.extend( hasValueType )

const _ArraySpecBase = z.object( {
    type: z.literal( 'array' )
} )

type _ArraySpecType = z.infer<typeof _ArraySpecBase> & {
    contains?: string | number | z.infer<typeof HashSpec>
}

const ArraySpec: z.ZodType<_ArraySpecType> = _ArraySpecBase.extend( {
    contains: z.lazy( () => z.union( [ z.string(), z.number(), HashSpec ] ).optional() )
} )

const MapSpec = _HashSpecBase.extend( {
    ...hasValueType,
    max_keys: z.number().optional()
} )
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by celestine-stripe** [reply]
This is indeed what I went with :) So it looks like I just need to accept the bit of duplication with `hasValueType`. Thanks!
